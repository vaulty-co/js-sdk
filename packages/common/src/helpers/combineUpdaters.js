/**
 * Combine updaters by their map
 * @param {Object<string, Function>}updaters
 * @returns {function(Object, Object): Object}
 */
const combineUpdaters = (updaters) => (prev = {}, next = {}) => (
  Object.keys(updaters).reduce(
    (result, updaterName) => {
      const combinedSettings = result;
      combinedSettings[updaterName] = updaters[updaterName](prev, next);
      return combinedSettings;
    },
    {},
  )
);

export default combineUpdaters;
export {
  combineUpdaters,
};
