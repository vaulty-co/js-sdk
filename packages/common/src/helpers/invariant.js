import invariant from 'invariant';

/**
 * Invariant, which is used as static class method
 * @param {boolean} condition
 * @param {string} message
 */
function staticInvariant(condition, message) {
  invariant(condition, `[${this.name}] ${message}`);
}

/**
 * Create invariant with prefixed message
 * @param {string} prefix
 */
function createInvariant(prefix) {
  /**
   * @param {boolean} condition
   * @param {string} message
   */
  return function prefixedInvariant(condition, message) {
    invariant(condition, `[${prefix}] ${message}`);
  };
}

export {
  staticInvariant,
  createInvariant,
};
