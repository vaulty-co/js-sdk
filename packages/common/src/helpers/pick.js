/**
 * Pick properties from object and create a new with them
 * @param {Object} [object = {}]
 * @param {Array} [properties = []]
 * @returns {Object}
 */
const pick = (object = {}, properties = []) => (
  Object.keys(object).reduce(
    (result, property) => {
      const pickedObject = result;
      if (properties.includes(property)) {
        pickedObject[property] = object[property];
      }
      return pickedObject;
    },
    {},
  )
);

export default pick;
export {
  pick,
};
