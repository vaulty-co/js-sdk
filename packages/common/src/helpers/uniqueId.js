import { v4 as uuid } from 'uuid';

/**
 * Generate unique ID with excluding collisions
 * @param {string} prefix
 * @returns {string}
 */
function uniqueId(prefix = '') {
  return `${prefix}${uuid()}`;
}

export default uniqueId;
export {
  uniqueId,
};
