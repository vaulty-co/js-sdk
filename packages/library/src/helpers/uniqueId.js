let counter = 0;

function uniqueId(prefix = '') {
  counter += 1;
  return `${prefix}${counter}`;
}

export default uniqueId;
export {
  uniqueId,
};
