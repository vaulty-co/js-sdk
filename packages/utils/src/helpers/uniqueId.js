import uuid from 'uuid/v4';

/**
 * Generate unique ID with excluding collisions
 * @param {string} prefix
 * @returns {string}
 */
function uniqueId(prefix = '') {
  return `${prefix}${uuid()}`;
}

export default uniqueId;
export {
  uniqueId,
};
