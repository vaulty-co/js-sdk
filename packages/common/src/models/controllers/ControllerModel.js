import { immerable } from 'immer';
import { uniqueId } from '@js-sdk/common/src/helpers/uniqueId';
import { pick } from '../../helpers/pick';

import {
  CONTROLLER_STATUSES,
  INITIAL_CONTROLLER_STATUS,
} from './constants';

const ALLOWED_MODEL_JSON_PROPERTIES = ['id', 'fieldsIds', 'status'];

/**
 * Controller's model
 * @class
 */
class ControllerModel {
  /**
   * Get controller statuses
   * @return {Object}
   * @constructor
   */
  static get STATUSES() {
    return CONTROLLER_STATUSES;
  }

  /**
   * Create ControllerModel by its JSON
   * @param {ControllerModelJSON} modelJSON
   */
  static of(modelJSON) {
    /**
     * @type {ControllerModelOptions}
     */
    const controllerModelOptions = pick(modelJSON, ALLOWED_MODEL_JSON_PROPERTIES);
    return new ControllerModel(controllerModelOptions);
  }

  /**
   * @param {ControllerModelOptions} options
   */
  constructor(options = {}) {
    const {
      fieldsIds = [],
    } = options;

    this[immerable] = true;

    /**
     * @type {string}
     */
    this.id = options.id || uniqueId();
    /**
     * @type {Array<string>}
     */
    this.fieldsIds = fieldsIds || [];
    /**
     * @type {ControllerStatus}
     */
    this.status = options.status || INITIAL_CONTROLLER_STATUS;
  }

  /**
   * Add fields to controller
   */
  addFields(fieldsIds) {
    this.fieldsIds = this.fieldsIds.concat(fieldsIds);
  }

  /**
   * Remove fields from controller
   */
  removeFields(fieldsIds) {
    this.fieldsIds = this.fieldsIds.filter((fieldId) => !fieldsIds.includes(fieldId));
  }

  /**
   * Set controller status
   * @param {ControllerStatus} status
   */
  setStatus(status) {
    this.status = {
      ...this.status,
      ...status,
    };
  }

  /**
   * Prepare a plain object for using in JSON.stringify()
   * @returns {ControllerModelJSON}
   */
  toJSON() {
    return pick(this, ALLOWED_MODEL_JSON_PROPERTIES);
  }
}

export default ControllerModel;
export {
  ControllerModel,
};
