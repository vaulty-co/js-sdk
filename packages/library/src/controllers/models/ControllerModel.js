import { immerable } from 'immer';

import { uniqueId } from '../../helpers/uniqueId';
import { FIELD_READINESS_STATUSES } from '../../fields/constants';

/**
 * ControllerModel statuses
 * @enum {string}
 */
const CONTROLLER_MODEL_STATUSES = {
  INITIALIZED: 'initialized',
  PROCESSING: 'processing',
  READY: 'ready',
};

/**
 * Controller's model
 * @class
 */
class ControllerModel {
  static get STATUSES() {
    return CONTROLLER_MODEL_STATUSES;
  }

  constructor(options = {}) {
    const {
      fieldsIds = [],
    } = options;

    this[immerable] = true;

    /**
     * @type {string}
     */
    this.id = uniqueId('form-');
    /**
     * @type {Array<string>}
     */
    this.fieldsIds = fieldsIds;
  }

  /**
   * Add fields to controller
   */
  addFields(fieldsIds) {
    this.fieldsIds = this.fieldsIds.concat(fieldsIds);
  }

  /**
   * Remove fields from controller
   */
  removeFields(fieldsIds) {
    this.fieldsIds = this.fieldsIds.filter((fieldId) => !fieldsIds.includes(fieldId));
  }

  // TODO - add memoization
  /**
   * Get fields status
   * @param {FieldsCollection} fieldsCollection
   * @return {CONTROLLER_MODEL_STATUSES}
   */
  getFieldsStatus(fieldsCollection) {
    const fields = this.fieldsIds.map((fieldId) => (
      fieldsCollection.getField(fieldId)
    ));
    const isReady = fields.every((field) => (
      field.status.readiness === FIELD_READINESS_STATUSES.READY
    ));
    return isReady ? CONTROLLER_MODEL_STATUSES.READY : CONTROLLER_MODEL_STATUSES.INITIALIZED;
  }

  /**
   * Set controller status
   * @param {CONTROLLER_MODEL_STATUSES} status
   */
  setStatus(status) {
    this.status = status;
  }
}

export default ControllerModel;
export {
  ControllerModel,
};
