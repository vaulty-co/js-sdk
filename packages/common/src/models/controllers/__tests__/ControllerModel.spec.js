import { uniqueId } from '@vaulty/common/src/helpers/uniqueId';

import {
  CONTROLLER_NODE_STATUSES,
  CONTROLLER_READINESS_STATUSES,
  CONTROLLER_VALIDATION_STATUSES,
} from '../constants';
import { ControllerModel } from '../ControllerModel';

jest.mock('@vaulty/common/src/helpers/uniqueId', () => {
  const { uniqueId: uniqueIdActual } = jest.requireActual('@vaulty/common/src/helpers/uniqueId');
  return {
    uniqueId: jest.fn(uniqueIdActual),
  };
});

describe('ControllerModel', () => {
  let controllerModel;

  beforeEach(() => {
    controllerModel = new ControllerModel({
      fieldsIds: ['field-1', 'field-2'],
    });
  });

  afterEach(() => {
    const { uniqueId: uniqueIdActual } = jest.requireActual('@vaulty/common/src/helpers/uniqueId');
    uniqueId.mockImplementation(uniqueIdActual);
  });

  describe('#constructor', () => {
    it('should create model with specified fieldsIds', () => {
      expect(controllerModel.fieldsIds).toEqual(['field-1', 'field-2']);
    });

    it('should set empty fieldsIds if it is not specified', () => {
      controllerModel = new ControllerModel();

      expect(controllerModel.fieldsIds).toEqual([]);
    });

    it('should setup initial status', () => {
      expect(controllerModel.status).toMatchSnapshot();
    });
  });

  describe('#addFields', () => {
    it('should add array of fields ids', () => {
      controllerModel.addFields(['field-3', 'field-4']);

      expect(controllerModel.fieldsIds).toEqual(['field-1', 'field-2', 'field-3', 'field-4']);
    });
  });

  describe('#removeFields', () => {
    it('should remove fields ids', () => {
      controllerModel.addFields(['field-3', 'field-4']);

      controllerModel.removeFields(['field-3', 'field-4']);

      expect(controllerModel.fieldsIds).toEqual(['field-1', 'field-2']);
    });
  });

  describe('#setStatus', () => {
    it('should set status with merging with previous', () => {
      controllerModel.setStatus({
        node: CONTROLLER_NODE_STATUSES.MOUNTED,
        validation: CONTROLLER_VALIDATION_STATUSES.VALID,
        readiness: CONTROLLER_READINESS_STATUSES.READY,
      });

      controllerModel.setStatus({
        validation: CONTROLLER_VALIDATION_STATUSES.INVALID,
      });

      expect(controllerModel.status).toMatchSnapshot();
    });
  });
});
