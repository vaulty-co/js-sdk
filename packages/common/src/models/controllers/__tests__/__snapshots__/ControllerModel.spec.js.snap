// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ControllerModel #constructor should setup initial status 1`] = `
Object {
  "node": "unmounted",
  "readiness": "loading",
}
`;

exports[`ControllerModel #setStatus should set status with merging with previous 1`] = `
Object {
  "node": "mounted",
  "readiness": "ready",
  "validation": "invalid",
}
`;
